common/persistence/nosql/nosqlplugin/cassandra/tasks_test.go

```diff
@@ -0,0 +1,1086 @@
+				query.EXPECT().WithContext(gomock.Any()).Return(query).Times(1)
+				query.EXPECT().MapScanCAS(gomock.Any()).DoAndReturn(func(prev map[string]interface{}) (bool, error) {
+					prev["range_id"] = int64(26)
+					return false, nil
+				}).Times(1)
+			},
+			wantErr: true,
+		},
+		{
+			name: "mapscan failed",
+			row: &nosqlplugin.TaskListRow{
+				DomainID:        "domain1",
+				TaskListName:    "tasklist1",
+				TaskListType:    1,
+				TaskListKind:    2,
+				AckLevel:        1000,
+				RangeID:         25,
+				LastUpdatedTime: ts,
+			},
+			queryMockFn: func(query *gocql.MockQuery) {
+				query.EXPECT().WithContext(gomock.Any()).Return(query).Times(1)
+				query.EXPECT().MapScanCAS(gomock.Any()).DoAndReturn(func(prev map[string]interface{}) (bool, error) {
+					return false, errors.New("some random error")
+				}).Times(1)
+			},
+			wantErr: true,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			query := gocql.NewMockQuery(ctrl)
+			tc.queryMockFn(query)
+			session := &fakeSession{
+				query: query,
+			}
+			client := gocql.NewMockClient(ctrl)
+			cfg := &config.NoSQL{}
+			logger := testlogger.New(t)
+			dc := &persistence.DynamicConfiguration{}
+
+			db := newCassandraDBFromSession(cfg, session, logger, dc, dbWithClient(client))
+
+			err := db.InsertTaskList(context.Background(), tc.row)
+
+			if (err != nil) != tc.wantErr {
+				t.Errorf("InsertTaskList() error = %v, wantErr %v", err, tc.wantErr)
+			}
+
+			if err != nil {
+				return
+			}
+
+			if diff := cmp.Diff(tc.wantQueries, session.queries); diff != "" {
+				t.Fatalf("Query mismatch (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestUpdateTaskList(t *testing.T) {
+	ts, err := time.Parse(time.RFC3339, "2024-04-01T22:08:41Z")
+	if err != nil {
+		t.Fatalf("Failed to parse time: %v", err)
+	}
+
+	tests := []struct {
+		name        string
+		row         *nosqlplugin.TaskListRow
+		prevRangeID int64
+		queryMockFn func(query *gocql.MockQuery)
+		wantQueries []string
+		wantErr     bool
+	}{
+		{
+			name:        "successfully applied",
+			prevRangeID: 25,
+			row: &nosqlplugin.TaskListRow{
+				DomainID:        "domain1",
+				TaskListName:    "tasklist1",
+				TaskListType:    1,
+				TaskListKind:    2,
+				AckLevel:        1000,
+				RangeID:         25,
+				LastUpdatedTime: ts,
+			},
+			queryMockFn: func(query *gocql.MockQuery) {
+				query.EXPECT().WithContext(gomock.Any()).Return(query).Times(1)
+				query.EXPECT().MapScanCAS(gomock.Any()).DoAndReturn(func(prev map[string]interface{}) (bool, error) {
+					return true, nil
+				}).Times(1)
+			},
+			wantQueries: []string{
+				`UPDATE tasks SET range_id = 25, task_list = {domain_id: domain1, name: tasklist1, type: 1, ack_level: 1000, kind: 2, last_updated: 2024-04-01T22:08:41Z } WHERE domain_id = domain1 and task_list_name = tasklist1 and task_list_type = 1 and type = 1 and task_id = -12345 IF range_id = 25`,
```

Comment:
having these document is really nice from a readability/documentation perspective